### PR TITLE
Fix remove parentheses bugs

### PR DESCRIPTION
**Pull request by Cosifne**
Related issue: [31103](https://github.com/dotnet/roslyn/issues/31103) and [32085](https://github.com/dotnet/roslyn/issues/32085)

Two problems of the current remove parentheses analyzer:
1. When conditional expression used as a left hand side value, 
`(x : ref a ? ref b) = c`
The parentheses should not be removed.
2. When conditional expression is used inside a string interpolation expression, and the whole string live inside an parentheses, like
`Func<bool, string> lambda = x => ($"{(x ? "foo" : "bar")}")`
The inner pair should not be removed.

### CODE REVIEW

File: src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs [hunk 1]
```diff
@@ -265,18 +273,17 @@ public static bool CanRemoveParentheses(this ParenthesizedExpressionSyntax node,
 
         private static bool RemovalMayIntroduceInterpolationAmbiguity(ParenthesizedExpressionSyntax node)
         {
-            // First, find the parenting interpolation. If we find a parenthesize expression first,
-            // we can bail out early.
+            // First, find the parenting interpolation.
+            // If we find a parenthesize expression first, it doesn't mean we can bail, we still need to check : or ::
+            // For example: Func<bool, string> lambda = x => ($"{(x ? "foo" : "bar")}"), the outer parenthesis expression 
+            // will let us wrongly think we can remove the inner parenthesis expression.
```
Comment (genlu) [inline src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs]: I think we can still bail early when a parenthesis is met *first*. Based on the original comment, it seems the issue is we keep iterating over the ancestors after encountering an interpolation. In the example you gave, we should have broken out the loop at `{(x ? "foo" : "bar")}`, so the enclosing `ParenthesizedExpression` should never be reached. Looks like someone has mistaken the `break` for the break of loop :)

File: src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs [hunk 2]
```diff
@@ -255,6 +255,14 @@ public static bool CanRemoveParentheses(this ParenthesizedExpressionSyntax node,
                 }
             }
 
+            // (condition ? ref a : ref b ) = SomeValue, parenthesis can't be removed for when conditional expression appears at left
+            // This syntax is only allowed since C# 7.2
+            if (expression.IsKind(SyntaxKind.ConditionalExpression) &&
+                node.IsLeftSideOfAnyAssignExpression())
+            {
+                return false;
+            }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs]: do we have tests that it *does* remove properly when on the rhs?

**Comment by genlu** [comment]
Also, please make commit messages more descriptive next time.

File: src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs (hunk 1 above)
Comment (Cosifne) [reply src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs]: make sense, I will update this part in next commit.

File: src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs (hunk 2 above)
Comment (Cosifne) [reply src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs]: I have updated the PR and give one.

**Comment by Cosifne** [comment]
Tag @vatsalyaagrawal @jinujoseph to approve

**Pull request merged**